File: src/python/experiments/run_repeated_experiments.py
import csv
import os

from src.python.backtrack.backtrack import hamiltonian_backtracking
from src.python.dynamic.dynamic import hamiltonian_dp
from src.python.heuristic.heuristic import hamiltonian_nearest_neighbor
from src.python.utils.decorators import ExecutionContext
from src.python.utils.load_graph import load_graph_txt


def run_repeated_experiments(base_dir='data',
                             results_filename='results/python/repeated_experiment_results.csv',
                             repetitions=5):
  results_data = []
  headers = [
      'instance_name', 'graph_type', 'size_n', 'algorithm', 'run_id',
      'time_seconds', 'cost', 'path_found', 'ops_count', 'memory_peak_kb'
  ]

  print(
      f'Iniciando experimento con {repetitions} repeticiones por instancia. Resultados en \'{results_filename}\''
  )

  if not os.path.exists(base_dir):
    print(f'Error: Directorio de instancias \'{base_dir}\' no encontrado.')
    return

  for graph_type in os.listdir(base_dir):
    type_path = os.path.join(base_dir, graph_type)
    if os.path.isdir(type_path):
      for filename in os.listdir(type_path):
        if filename.endswith('.txt'):
          filepath = os.path.join(type_path, filename)
          print(f'\n--- Probando instancia: {filepath} ---')

          try:
            graph_matrix = load_graph_txt(filepath)
          except Exception as e:
            print(f'  -> Error al cargar el grafo: {e}')
            continue

          n_size = len(graph_matrix)

          algorithms = {
              'Backtracking': hamiltonian_backtracking,
              'DP': hamiltonian_dp,
              'NearestNeighbor': hamiltonian_nearest_neighbor
          }

          for algo_name, algo_func in algorithms.items():
            if algo_name == 'DP' and n_size > 21:
              continue
            for run_id in range(1, repetitions + 1):
              context = ExecutionContext()
              path, cost = algo_func(graph_matrix, context=context)
              duration = context.time_seconds
              ops_count = context.ops_count
              mem_peak = context.memory_peak_kb
              path_found = 1 if path else 0
              cost_for_csv = cost if path else -1

              results_data.append([
                  filename, graph_type, n_size, algo_name, run_id, duration,
                  cost_for_csv, path_found, ops_count, mem_peak
              ])

              print(
                  f'  -> {algo_name} [Repetición {run_id}]: {'Éxito' if path_found else 'Fallo'}, '
                  f'Costo: {cost_for_csv:.1f}, '
                  f'Tiempo: {duration:.4f}s, '
                  f'Memoria: {mem_peak:.2f} KB, '
                  f'Ops: {ops_count:,}')

  try:
    with open(results_filename, 'w', newline='') as f:
      writer = csv.writer(f)
      writer.writerow(headers)
      writer.writerows(results_data)
    print(
        f'\n¡Experimento repetido completado! Resultados guardados en \'{results_filename}\''
    )
  except IOError:
    print(f'\nError: No se pudo escribir en el archivo \'{results_filename}\'.')


if __name__ == '__main__':
  run_repeated_experiments(base_dir='data',
                           results_filename='results/python/repeated_results.csv',
                           repetitions=5)

File: src/lua/experiments/run_repeated_experiments.lua
local bt = require('src.lua.backtrack.backtrack')
local dp = require('src.lua.dynamic.dynamic')
local nn = require('src.lua.heuristic.heuristic')
local loader = require('src.lua.utils.load_graph')
local decorators = require('src.lua.utils.decorators')

local results_filename = 'results/lua/repeated_experiment_results.csv'
local base_dir = 'data'
local repetitions = 5

-- Función para obtener todos los archivos .txt en subdirectorios de base_dir (solo para sistemas UNIX-like)
local function get_files_in_dir(base_path)
  local handle = io.popen('find ' .. base_path .. ' -type f -name "*.txt"')
  if not handle then
    error('No se pudo abrir el proceso para listar archivos en: ' .. base_path)
  end

  local files = {}
  for file in handle:lines() do
    table.insert(files, file)
  end
  handle:close()
  return files
end

-- Crear directorio si no existe
local function ensure_dir_exists(path)
  os.execute('mkdir -p ' .. path)
end

ensure_dir_exists('results/lua')

-- Cabecera CSV
local headers = {
  'instance_name',
  'graph_type',
  'size_n',
  'algorithm',
  'run_id',
  'time_seconds',
  'cost',
  'path_found',
  'ops_count',
  'memory_peak_kb',
}

-- Escribir resultados al CSV
local function write_csv(path, headers, data)
  local file = io.open(path, 'w')
  if not file then
    error('No se pudo escribir el archivo de resultados: ' .. path)
  end
  file:write(table.concat(headers, ',') .. '\n')
  for _, row in ipairs(data) do
    local line = {}
    for _, val in ipairs(row) do
      table.insert(line, tostring(val))
    end
    file:write(table.concat(line, ',') .. '\n')
  end
  file:close()
end

-- Ejecutar experimento
local function run_experiments()
  local results = {}
  print('Iniciando experimento repetido...')

  local algorithms = {
    Backtracking = bt,
    DP = dp,
    NearestNeighbor = nn,
  }

  for _, filepath in ipairs(get_files_in_dir(base_dir)) do
    print('\n--- Probando instancia: ' .. filepath .. ' ---')

    local graph = loader.load_graph_txt(filepath)
    local n = #graph
    local filename = filepath:match('([^/]+)%.txt$') or 'unknown'
    local graph_type = filepath:match(base_dir .. '/([^/]+)/') or 'unknown'

    for algo_name, algo_func in pairs(algorithms) do
      if algo_name == 'DP' and n > 21 then
        print('  -> DP omitido por tamaño > 21')
      else
        for run_id = 1, repetitions do
          local context = decorators.new_execution_context()
          local path, cost = algo_func(graph, context)

          local time = context.time_seconds or 0
          local mem = context.memory_peak_kb or 0
          local ops = context.ops_count or 0
          local found = path and 1 or 0
          local real_cost = path and cost or -1

          table.insert(results, {
            filename,
            graph_type,
            n,
            algo_name,
            run_id,
            string.format('%.6f', time),
            real_cost,
            found,
            ops,
            string.format('%.2f', mem),
          })

          print(string.format(
            '  -> %s [Repetición %d]: %s | Costo: %s | Tiempo: %.4fs | Memoria: %.2f KB | Ops: %d',
            algo_name, run_id, found == 1 and 'Éxito' or 'Fallo', real_cost, time,
            mem, ops
          ))
        end
      end
    end
  end

  write_csv(results_filename, headers, results)
  print('\n¡Experimento completado! Resultados guardados en ' .. results_filename)
end

run_experiments()
